src/bridge/cockpitcgroupsamples.c

```diff
@@ -21,119 +21,223 @@
+  val = read_int64 (dirfd, cgroup, "memory.usage_in_bytes");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.usage", cgroup, val);
 
-      cockpit_samples_sample (samples, "cgroup.cpu.usage", cgroup, cpuacct_usage/1000000);
-      cockpit_samples_sample (samples, "cgroup.cpu.shares", cgroup, cpu_shares);
-    }
+  val = read_int64 (dirfd, cgroup, "memory.limit_in_bytes");
+  /* If at max for arch, then unlimited => zero */
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.limit", cgroup, val);
+
+  val = read_int64 (dirfd, cgroup, "memory.memsw.usage_in_bytes");
+  if (val >= 0 && val < G_MAXINT64)
+      cockpit_samples_sample (samples, "cgroup.memory.sw-usage", cgroup, val);
+
+  val = read_int64 (dirfd, cgroup, "memory.memsw.limit_in_bytes");
+  /* If at max for arch, then unlimited => zero */
+  if (val > 0 && val < G_MAXINT64)
+      cockpit_samples_sample (samples, "cgroup.memory.sw-limit", cgroup, val);
+}
+
+static void
+collect_cpu_v1 (CockpitSamples *samples,
+                int dirfd,
+                const char *cgroup)
+{
+  gint64 val;
+
+  val = read_int64 (dirfd, cgroup, "cpuacct.usage");
+  if (val >= 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.cpu.usage", cgroup, val/1000000);
+
+  val = read_int64 (dirfd, cgroup, "cpu.shares");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.cpu.shares", cgroup, val);
+}
+
+static void
+collect_v2 (CockpitSamples *samples,
+            int dirfd,
+            const char *cgroup)
+{
+  gint64 val;
+
+  /* memory.current: single unsigned value in bytes */
+  val = read_int64 (dirfd, cgroup, "memory.current");
+  if (val >= 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.usage", cgroup, val);
+
+  /* memory.max: literally says "max" if there is no limit set, which ends up as "0" after integer conversion;
+   * only create samples for actually limited cgroups */
+  val = read_int64 (dirfd, cgroup, "memory.max");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.limit", cgroup, val);
+
+  /* same as above for swap */
+  val = read_int64 (dirfd, cgroup, "memory.swap.current");
+  if (val >= 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.sw-usage", cgroup, val);
+
+  val = read_int64 (dirfd, cgroup, "memory.swap.max");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.sw-limit", cgroup, val);
+
+  /* cpu.weight: only exists if cpu controller is enabled; integer in range [1, 10000] */
+  val = read_int64 (dirfd, cgroup, "cpu.weight");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.cpu.shares", cgroup, val);
+
+  /* cpu.stat: keyed file:
+     usage_usec 50000
+     user_usec 40000
+     system_usec 10000
+     */
+  val = read_prefixed_int64 (dirfd, cgroup, "cpu.stat", "usage_usec ");
```

Comment:
Your approach here gives me pause.  What if some day they add a field called `max_usage_usec`?  Then this would match on that...

This is why I was leaning toward iterating over the lines...  I guess another thing you could do is to ensure that any match is either at the start of the buffer, or proceeded by a `\n` (and search for other matches, if not).

What do you think about this issue?

Reply:
Changed to checking the preceding char now, and renamed the function to match what it does.